jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/MessageHandler.java

```diff
@@ -200,29 +200,29 @@ protected void onTextFrame(Frame frame, Callback callback)
                 long currentSize = frame.getPayload().remaining() + textBuffer.length();
                 if (currentSize > maxSize)
                     throw new MessageTooLargeException("Message larger than " + maxSize + " bytes");
-
                 textBuffer.append(frame.getPayload());
             }
 
             if (frame.isFin())
             {
-                onText(textBuffer.toString(), callback);
+                onText(textBuffer.takeString(CharacterCodingException::new), callback);
                 textBuffer.reset();
             }
             else
             {
+                if (textBuffer.hasCodingErrors())
+                    throw new CharacterCodingException();
                 callback.succeeded();
             }
         }
-        catch (Utf8Appendable.NotUtf8Exception e)
+        catch (CharacterCodingException e)
```

Comment:
Remove this entire catch block (once you change to throwing a `BadPayloadException`)